File: mcp_server/demo/fixtures.py
"""Canned mailbox data for the no-auth demo mount.

Python port of ``mcp_server/dev_preview/src/fixtures.ts`` - the same fictional
inbox the MCP-UI dev preview renders, expressed as real service output models
so the demo tools' wire format (structuredContent, outputSchema) is identical
to production. When a tool contract changes, update both files.

Every address and company here is fictional; the dataset is deliberately
recognizable as a demo (see the account owner ``you@startup.com``).
"""

from datetime import UTC, datetime

from models.gmail import (
    GmailAttachment,
    GmailCuratedThread,
    GmailCurateInboxResult,
    GmailDraft,
    GmailLabelChip,
    GmailThread,
    GmailThreadMessage,
)

DEMO_USER = "demo"
DEMO_OWNER_ADDRESS = "you@startup.com"

# A minimal but valid single-page PDF, standing in for the term-sheet
# attachment so the inbox reader's download/preview path works in the demo.
DEMO_PDF_BASE64 = (
    "JVBERi0xLjEKMSAwIG9iajw8L1R5cGUvQ2F0YWxvZy9QYWdlcyAyIDAgUj4+ZW5kb2JqCjIg"
    "MCBvYmo8PC9UeXBlL1BhZ2VzL0tpZHNbMyAwIFJdL0NvdW50IDE+PmVuZG9iagozIDAgb2Jq"
    "PDwvVHlwZS9QYWdlL1BhcmVudCAyIDAgUi9NZWRpYUJveFswIDAgMjAwIDIwMF0+PmVuZG9i"
    "agp0cmFpbGVyPDwvUm9vdCAxIDAgUj4+"
)


def _dt(value: str) -> datetime:
    return datetime.fromisoformat(value).replace(tzinfo=UTC)


def _curated(sender: str, **kw) -> GmailCuratedThread:
    """Build a curated thread.

    ``from`` is an alias-only field (Python keyword), so construction goes
    through ``model_validate`` - the same pattern as the curate service.
    """
    return GmailCuratedThread.model_validate({"from": sender, **kw})


def _message(sender: str, **kw) -> GmailThreadMessage:
    """Build a thread message with the aliased ``from`` field."""
    return GmailThreadMessage.model_validate({"from": sender, **kw})


_UNREAD = GmailLabelChip(name="Unread", bg_color="#e8f0fe", text_color="#1a73e8")

CURATED_THREADS: list[GmailCuratedThread] = [
    _curated(
        "Dana Whitfield <dana@northwind.vc>",
        thread_id="t-1001",
        subject="Series A term sheet - final redlines",
        snippet=(
            "Great call today. Attaching the final redlines - one open point on "
            "the liquidation preference…"
        ),
        last_message_at=_dt("2026-07-05T09:14:00"),
        importance_score=0.97,
        reasons=["Unread", "VIP sender", "Awaiting reply"],
        labels=[
            _UNREAD,
            GmailLabelChip(name="Finance", bg_color="#e6f4ea", text_color="#137333"),
        ],
        has_draft=False,
    ),
    _curated(
        "Priya Nair <priya@peoplehq.io>",
        thread_id="t-1002",
        subject="Re: Onsite interview loop for Staff Eng",
        snippet=(
            "Confirming Thursday 10am-2pm. I've looped in the panel. Let me know "
            "if the schedule works…"
        ),
        last_message_at=_dt("2026-07-05T07:41:00"),
        importance_score=0.82,
        reasons=["Unread", "Calendar"],
        labels=[_UNREAD],
        has_draft=True,
        draft_id="d-9001",
    ),
    _curated(
        "billing@vercel.com",
        thread_id="t-1003",
        subject="Your invoice #INV-2043 is ready",
        snippet="Your monthly invoice is now available. Amount due: $240.00…",
        last_message_at=_dt("2026-07-04T22:03:00"),
        importance_score=0.44,
        reasons=["Receipt"],
        labels=[
            GmailLabelChip(name="Receipts", bg_color="#fef7e0", text_color="#b06000")
        ],
        has_draft=False,
    ),
]

# Seed draft on t-1002 - also the draft the composer app opens by default.
SEED_DRAFT = GmailDraft(
    draft_id="d-9001",
    to="priya@peoplehq.io",
    cc="",
    bcc="",
    subject="Re: Onsite interview loop for Staff Eng",
    body=(
        "Hi Priya,\n\nThursday 10am-2pm works great - please send the calendar "
        "holds and I'll confirm with the panel on my side.\n\nLooking forward "
        "to it.\n\nBest,\nAlex"
    ),
    thread_id="t-1002",
    attachments=[],
)

THREADS: dict[str, GmailThread] = {
    "t-1001": GmailThread(
        thread_id="t-1001",
        messages=[
            _message(
                "Dana Whitfield <dana@northwind.vc>",
                message_id="m-1",
                to=DEMO_OWNER_ADDRESS,
                date=_dt("2026-07-05T09:14:00"),
                subject="Series A term sheet - final redlines",
                body_html=(
                    "<p>Hi,</p><p>Great call today. Attaching the final redlines"
                    " - one open point on the <b>liquidation preference</b> "
                    "(we're proposing 1x non-participating).</p><p>If that "
                    "works, we can sign this week.</p><p>Best,<br/>Dana</p>"
                ),
                attachments=[
                    GmailAttachment(
                        filename="termsheet-v7.pdf",
                        mime_type="application/pdf",
                        size=184320,
                        attachment_id="att-termsheet",
                    )
                ],
            )
        ],
    ),
    "t-1002": GmailThread(
        thread_id="t-1002",
        messages=[
            _message(
                "Priya Nair <priya@peoplehq.io>",
                message_id="m-2",
                to=DEMO_OWNER_ADDRESS,
                date=_dt("2026-07-05T07:41:00"),
                subject="Re: Onsite interview loop for Staff Eng",
                body_text=(
                    "Confirming Thursday 10am-2pm. I've looped in the panel. "
                    "Let me know if the schedule works and I'll send calendar "
                    "holds.\n\nThanks,\nPriya"
                ),
                attachments=[],
            )
        ],
        draft=SEED_DRAFT,
    ),
    "t-1003": GmailThread(
        thread_id="t-1003",
        messages=[
            _message(
                "billing@vercel.com",
                message_id="m-3",
                to=DEMO_OWNER_ADDRESS,
                date=_dt("2026-07-04T22:03:00"),
                subject="Your invoice #INV-2043 is ready",
                body_text=(
                    "Your monthly invoice is now available. Amount due: "
                    "$240.00. No action needed - auto-pay is on."
                ),
                attachments=[],
            )
        ],
    ),
}


def curated_inbox(limit: int = 10) -> GmailCurateInboxResult:
    """The canned ranked inbox, truncated to ``limit``."""
    return GmailCurateInboxResult(threads=CURATED_THREADS[:limit])


def get_thread(thread_id: str) -> GmailThread:
    """Look up a fixture thread; raise a demo-friendly error for unknown ids."""
    thread = THREADS.get(thread_id)
    if thread is None:
        known = ", ".join(sorted(THREADS))
        raise ValueError(
            f"Unknown demo thread {thread_id!r}. This demo mailbox has three "
            f"threads: {known}. Call gmail_curate_inbox to list them."
        )
    return thread.model_copy(deep=True)

File: mcp_server/dev_preview/src/fixtures.ts
// Fixture data for the MCP-UI dev preview. No Gmail, no OAuth, no network -
// these canned payloads stand in for the responses a real MCP server would
// return, so a committed app bundle can be rendered in isolation.
//
// Shapes mirror the app types (see mcp_server/apps/*/src/*.tsx). When a tool's
// contract changes, update the matching entry here.

type ToolResult = {
  content: { type: "text"; text: string }[];
  structuredContent: Record<string, unknown>;
};

export function ok(data: unknown): ToolResult {
  return {
    content: [{ type: "text", text: JSON.stringify(data) }],
    structuredContent: (data ?? {}) as Record<string, unknown>,
  };
}

// --- gmail_inbox ----------------------------------------------------------

const INBOX_CURATE = {
  threads: [
    {
      thread_id: "t-1001",
      subject: "Series A term sheet - final redlines",
      from: "Dana Whitfield <dana@northwind.vc>",
      snippet:
        "Great call today. Attaching the final redlines - one open point on the liquidation preference…",
      last_message_at: "2026-07-05T09:14:00Z",
      importance_score: 0.97,
      reasons: ["Unread", "VIP sender", "Awaiting reply"],
      labels: [
        { name: "Unread", bg_color: "#e8f0fe", text_color: "#1a73e8" },
        { name: "Finance", bg_color: "#e6f4ea", text_color: "#137333" },
      ],
      has_draft: false,
    },
    {
      thread_id: "t-1002",
      subject: "Re: Onsite interview loop for Staff Eng",
      from: "Priya Nair <priya@peoplehq.io>",
      snippet:
        "Confirming Thursday 10am–2pm. I've looped in the panel. Let me know if the schedule works…",
      last_message_at: "2026-07-05T07:41:00Z",
      importance_score: 0.82,
      reasons: ["Unread", "Calendar"],
      labels: [{ name: "Unread", bg_color: "#e8f0fe", text_color: "#1a73e8" }],
      has_draft: true,
      draft_id: "d-9001",
    },
    {
      thread_id: "t-1003",
      subject: "Your invoice #INV-2043 is ready",
      from: "billing@vercel.com",
      snippet: "Your monthly invoice is now available. Amount due: $240.00…",
      last_message_at: "2026-07-04T22:03:00Z",
      importance_score: 0.44,
      reasons: ["Receipt"],
      labels: [{ name: "Receipts", bg_color: "#fef7e0", text_color: "#b06000" }],
      has_draft: false,
    },
  ],
};

const THREADS: Record<string, unknown> = {
  "t-1001": {
    thread_id: "t-1001",
    messages: [
      {
        message_id: "m-1",
        from: "Dana Whitfield <dana@northwind.vc>",
        to: "you@startup.com",
        date: "2026-07-05T09:14:00Z",
        subject: "Series A term sheet - final redlines",
        body_html:
          // Remote img mirrors services/_gmail_fake_backend.py: exercises the
          // blocked-by-default "Show images" banner (.invalid never resolves).
          "<p>Hi,</p><p>Great call today. Attaching the final redlines - one open point on the <b>liquidation preference</b> (we're proposing 1x non-participating).</p><p>If that works, we can sign this week.</p><p>Best,<br/>Dana</p><img src=\"https://img.invalid/northwind-logo.png\" alt=\"Northwind Ventures\">",
        attachments: [
          { filename: "termsheet-v7.pdf", mime_type: "application/pdf", size: 184320 },
        ],
      },
    ],
  },
  "t-1002": {
    thread_id: "t-1002",
    messages: [
      {
        message_id: "m-2",
        from: "Priya Nair <priya@peoplehq.io>",
        to: "you@startup.com",
        date: "2026-07-05T07:41:00Z",
        subject: "Re: Onsite interview loop for Staff Eng",
        body_text:
          "Confirming Thursday 10am-2pm. I've looped in the panel. Let me know if the schedule works and I'll send calendar holds.\n\nThanks,\nPriya",
        attachments: [],
      },
    ],
    draft: {
      draft_id: "d-9001",
      to: "priya@peoplehq.io",
      subject: "Re: Onsite interview loop for Staff Eng",
      body: "Thursday works great - please send the holds. Looking forward to it.",
      thread_id: "t-1002",
      attachments: [],
    },
  },
  "t-1003": {
    thread_id: "t-1003",
    messages: [
      {
        message_id: "m-3",
        from: "billing@vercel.com",
        to: "you@startup.com",
        date: "2026-07-04T22:03:00Z",
        subject: "Your invoice #INV-2043 is ready",
        body_text:
          "Your monthly invoice is now available. Amount due: $240.00. No action needed - auto-pay is on.",
        attachments: [],
      },
    ],
  },
};

// --- gmail_composer -------------------------------------------------------

const COMPOSER_DRAFT = {
  draft_id: "d-9001",
  to: "priya@peoplehq.io",
  cc: "",
  bcc: "",
  subject: "Re: Onsite interview loop for Staff Eng",
  body:
    "Hi Priya,\n\nThursday 10am–2pm works great - please send the calendar holds and I'll confirm with the panel on my side.\n\nLooking forward to it.\n\nBest,\nAlex",
  thread_id: "t-1002",
  attachments: [],
};

// --- dispatch -------------------------------------------------------------

/** The first tool result pushed to the app on init (drives the initial paint). */
// --- pdf_signer -----------------------------------------------------------

// A real (tiny) flat NDA, filled via the actual pdf_edit engine, so pdf.js
// renders authentic bytes. Regenerate with tests/pdf_fixtures.make_flat_pdf +
// services.pdf_edit_engine.apply_ops if the fixture builders change.
const NDA_FILLED_B64 =
  "JVBERi0xLjMKJeLjz9MKMSAwIG9iago8PAovVHlwZSAvQ2F0YWxvZwovUGFnZXMgMiAwIFIKPj4KZW5kb2JqCjIgMCBvYmoKPDwKL1R5cGUgL1BhZ2VzCi9Db3VudCAxCi9LaWRzIFsgMyAwIFIgXQo+PgplbmRvYmoKMyAwIG9iago8PAovVHlwZSAvUGFnZQovUmVzb3VyY2VzIDw8Ci9Gb250IDw8Ci9GMSA0IDAgUgovUGRmRWRpdEYxIDcgMCBSCj4+Cj4+Ci9NZWRpYUJveCBbIDAuMCAwLjAgNjEyIDc5MiBdCi9QYXJlbnQgMiAwIFIKL0NvbnRlbnRzIDUgMCBSCj4+CmVuZG9iago0IDAgb2JqCjw8Ci9UeXBlIC9Gb250Ci9TdWJ0eXBlIC9UeXBlMQovQmFzZUZvbnQgL0hlbHZldGljYQo+PgplbmRvYmoKNSAwIG9iagpbIDggMCBSIDkgMCBSIF0KZW5kb2JqCjYgMCBvYmoKPDwKL1Byb2R1Y2VyIChweXBkZikKPj4KZW5kb2JqCjcgMCBvYmoKPDwKL1R5cGUgL0ZvbnQKL1N1YnR5cGUgL1R5cGUxCi9CYXNlRm9udCAvSGVsdmV0aWNhCj4+CmVuZG9iago4IDAgb2JqCjw8Ci9MZW5ndGggMTc3Cj4+CnN0cmVhbQpxCkJUIC9GMSAxNCBUZiA3MiA3MDAgVGQgKE5PTi1ESVNDTE9TVVJFIEFHUkVFTUVOVCkgVGogRVQKQlQgL0YxIDEwIFRmIDcyIDY1MCBUZCAoTmFtZTopIFRqIEVUCkJUIC9GMSAxMCBUZiA3MiA2MDAgVGQgKERhdGU6KSBUaiBFVApCVCAvRjEgMTAgVGYgNzIgNTUwIFRkIChTaWduYXR1cmU6KSBUaiBFVAoKUQoKZW5kc3RyZWFtCmVuZG9iago5IDAgb2JqCjw8Ci9MZW5ndGggMTM2Cj4+CnN0cmVhbQpxCjAuMCAwLjAgNjEyIDc5MiByZQpXCm4KQlQKL1BkZkVkaXRGMSAxMCBUZgoxMTAgNjUwIFRkCihFaXRvIE1peWFtdXJhKSBUagpFVApCVAovUGRmRWRpdEYxIDEwIFRmCjEwNSA2MDAgVGQKKDIwMjZcMDU1MDdcMDU1MjIpIFRqCkVUClEKCmVuZHN0cmVhbQplbmRvYmoKeHJlZgowIDEwCjAwMDAwMDAwMDAgNjU1MzUgZiAKMDAwMDAwMDAxNSAwMDAwMCBuIAowMDAwMDAwMDY0IDAwMDAwIG4gCjAwMDAwMDAxMjMgMDAwMDAgbiAKMDAwMDAwMDI3MiAwMDAwMCBuIAowMDAwMDAwMzQyIDAwMDAwIG4gCjAwMDAwMDAzNzMgMDAwMDAgbiAKMDAwMDAwMDQxMiAwMDAwMCBuIAowMDAwMDAwNDgyIDAwMDAwIG4gCjAwMDAwMDA3MTAgMDAwMDAgbiAKdHJhaWxlcgo8PAovU2l6ZSAxMAovUm9vdCAxIDAgUgovSW5mbyA2IDAgUgo+PgpzdGFydHhyZWYKODk3CiUlRU9GCg==";

const PDF_SIGNER_REQUEST = {
  doc_id: "doc-fixture-nda",
  status: "awaiting_user_signature",
  guidance: "The document is now locked for signing.",
};

const PDF_SIGNER_DOC = {
  doc_id: "doc-fixture-nda",
  filename: "nda.pdf",
  status: "awaiting_signature",
  page_count: 1,
  stamp_page: 1,
  // From services.pdf_signing.resolve_stamp_rect for placement (1, 140, 545).
  stamp_rect: [140.0, 541.0, 340.0, 577.0],
  data_base64: NDA_FILLED_B64,
};

export function initialResult(app: string): ToolResult {
  if (app === "gmail_composer") return ok(COMPOSER_DRAFT);
  if (app === "pdf_signer") return ok(PDF_SIGNER_REQUEST);
  return ok(INBOX_CURATE);
}

/** Answer an app's callServerTool by name, as a real server would. */
export function dispatch(name: string, args: Record<string, unknown>): ToolResult {
  switch (name) {
    case "gmail_inbox.refresh":
      return ok(INBOX_CURATE);
    case "gmail_inbox.open_thread":
    case "gmail_composer.get_thread":
      return ok(THREADS[String(args.thread_id)] ?? {});
    case "gmail_composer.refresh":
      return ok(COMPOSER_DRAFT);
    case "pdf_signer.get_document":
      return ok(PDF_SIGNER_DOC);
    case "pdf_signer.sign":
      return ok({
        doc_id: "doc-fixture-nda",
        status: "signed",
        signed_by: String(args.typed_name ?? "Signer"),
        signed_at_utc: "2026-07-22T12:00:00+00:00",
        message: `Signed by ${args.typed_name} on 2026-07-22T12:00:00+00:00.`,
      });
    case "pdf_signer.cancel":
      return ok({ doc_id: "doc-fixture-nda", status: "open" });
    case "gmail_composer.send":
      // A real message_id so the composer's Sent state (and the model-context
      // push it triggers) carries a plausible identifier.
      return ok({ message_id: "msg-fixture-0042", thread_id: "t-1002" });
    default:
      // set_focus, mark_read, archive, mark_done, reply, forward, save_draft,
      // send, discard … acknowledged with an empty structured result.
      return ok({});
  }
}
